pandar_pointcloud/src/lib/decoder/pandar40_decoder.cpp

```diff
@@ -86,81 +94,100 @@ void Pandar40Decoder::unpack(const pandar_msgs::PandarPacket& raw_packet)
   return;
 }
 
-PointcloudXYZIRADT Pandar40Decoder::convert(int block_id)
+PointXYZIRADT Pandar40Decoder::build_point(int block_id, int unit_id, int8_t return_type)
 {
-  PointcloudXYZIRADT block_pc(new pcl::PointCloud<PointXYZIRADT>);
+  const auto& block = packet_.blocks[block_id];
+  const auto& unit = block.units[unit_id];
+  double unix_second = static_cast<double>(timegm(&packet_.t));
+  PointXYZIRADT point;
 
-  // double unix_second = raw_packet.header.stamp.toSec() // system-time (packet receive time)
-  double unix_second = static_cast<double>(timegm(&packet_.t));  // sensor-time (ppt/gps)
+  double xyDistance = unit.distance * cosf(deg2rad(elev_angle_[unit_id]));
 
-  for (auto unit_id : firing_order_) {
-    PointXYZIRADT point;
-    const auto& block = packet_.blocks[block_id];
-    const auto& unit = block.units[unit_id];
-    // skip invalid points
-    if (unit.distance <= 0.1 || unit.distance > 200.0) {
-      continue;
-    }
-    double xyDistance = unit.distance * cosf(deg2rad(elev_angle_[unit_id]));
+  point.x = static_cast<float>(
+      xyDistance * sinf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
+  point.y = static_cast<float>(
+      xyDistance * cosf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
+  point.z = static_cast<float>(unit.distance * sinf(deg2rad(elev_angle_[unit_id])));
 
-    point.x = static_cast<float>(
-        xyDistance * sinf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
-    point.y = static_cast<float>(
-        xyDistance * cosf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
-    point.z = static_cast<float>(unit.distance * sinf(deg2rad(elev_angle_[unit_id])));
+  point.intensity = unit.intensity;
+  point.distance = unit.distance;
+  point.ring = unit_id;
+  point.azimuth = block.azimuth + round(azimuth_offset_[unit_id] * 100.0f);
+  point.return_type = return_type;
+  point.time_stamp = unix_second + (static_cast<double>(packet_.usec)) / 1000000.0;
+  point.time_stamp -= (static_cast<double>(block_offset_dual_[block_id] + firing_offset_[unit_id]) / 1000000.0f);
```

Comment:
It's a good idea to use the same `build_point` for convert() and convert_dual(), but the time offset table is different between single and dual mode. 
https://github.com/tier4/hesai_pandar/blob/tier4/master/pandar_pointcloud/src/lib/decoder/pandar40_decoder.cpp#L25-L28

Reply:
Well spotted, I will fix this!